examples/simple-grep/README.md

```diff
@@ -0,0 +1,129 @@
+import qualified Iris
+import qualified Paths_iris as Autogen
+
+newtype App a = App
+    { unApp :: Iris.CliApp SimpleGrep () a
+    } deriving newtype
+        ( Functor
+        , Applicative
+        , Monad
+        , MonadIO
+        , MonadReader (Iris.CliEnv SimpleGrep ())
+        )
+```
+
+It uses `Iris.CliApp SimpleGrep () a` that means `CliApp` will consume `SimpleGrep` data of commands for CLI description. Also, there's `()` in the second argument, it could be a specific environment type for our application, but in our case, it isn't necessary.
+`SimpleGrep` is described below:
+
+```haskell
+data SimpleGrep = SimpleGrep
+    {
+      filePath :: !String
+    , substring :: !String
+    }
+```
+
+After the creation of a base monad with data-type of commands, let's instantiate settings to Iris:
+
+```haskell
+appSettings :: Iris.CliEnvSettings SimpleGrep ()
+appSettings = Iris.defaultCliEnvSettings
+    { Iris.cliEnvSettingsHeaderDesc = "Iris usage example"
+    , Iris.cliEnvSettingsProgDesc = "A simple grep utility"
+    , Iris.cliEnvSettingsVersionSettings =
+        Just (Iris.defaultVersionSettings Autogen.version)
+            { Iris.versionSettingsMkDesc = \v -> "Simple grep utility v" <> v
+            }
+    , Iris.cliEnvSettingsRequiredTools = []
+
+    , Iris.cliEnvSettingsCmdParser = SimpleGrep
+        <$> Opt.strOption
+            (  Opt.long "filePath"
+            <> Opt.short 'f'
+            <> Opt.metavar "PATH"
+            <> Opt.help "filePath to find and grep file"
+            )
+        <*> Opt.strOption
+            (  Opt.long "substring"
+            <> Opt.short 's'
+            <> Opt.help "Substring to find and highlight"
+            )
+    }
+```
+
+We describe settings for the app based on `defaultCliEnvSettings`. It helps when we don't need some option to describe, so we can just skip it. Here we can write a CLI-program description, put required tools with a list and then pass commands with the `Parser a` type from `optparse-applicative` to the `cliEnvSettingsCmdParser` field.
+
+Finally, there's some little boilerplate code for toy `grep` reimplementation, see the remaining part of the example:
+
+```haskell
+
+occurences :: [T.Text] -> T.Text -> [(Int, T.Text)]
+occurences lines substring = go lines 0
+    where
+        go (x:xs) idx | substring `T.isInfixOf` x = (idx, x) : go xs (idx + 1)
+                      | otherwise = go xs (idx + 1)
+        go [] _ = []
+
+occurencesPrinter :: [(Int, T.Text)] -> App ()
+occurencesPrinter = mapM_ unpack
+    where
+        unpack :: (Int, T.Text) -> App ()
+        unpack (idx, line) = do
+            liftIO $ printIdx idx
+            liftIO $ putStr ":"
+            printLine line
+
+        printIdx idx = putStr $ " " ++ show idx
+
+        printLine :: T.Text -> App ()
+        printLine x = Iris.putStdoutColouredLn
+            (Colourista.formatWith [Colourista.yellow, Colourista.bold])
+                $ BSL.toStrict $ TLE.encodeUtf8 x
```

Comment:
I propose a slightly different change:

Let's write the number and `:` to `stderr` using `putStderrColouredLn` and the rest with just `putStdoutColouredLn` and no colouring. This way, if you redirect `simple-grep` output to another tool, you'll get only text lines.

Reply:
It would be great, and I did it. But, it's okay that output there now like this structure? There is a new line after the number and `:`. It cause of `putStdoutColoured` or `putStderrColoured` doesn't exists. Maybe if it's not okay, there exists a solution for that. I'm not so well familiar with the terminal's escaping codes :raised_hands:.

```shell
cabal exec simple-grep -- -f /some/dir/iris/iris.cabal -s iris

Starting grepping 🔥 

 file name: iris.cabal 
 2:
name:                iris
 7:
    See [README.md](https://github.com/chshersh/iris#iris) for more details.
 8:
homepage:            https://github.com/chshersh/iris
 9:
bug-reports:         https://github.com/chshersh/iris/issues
 26:
  location:            https://github.com/chshersh/iris.git
 79:
  build-depends:       , iris
 119:
  autogen-modules:     Paths_iris
 120:
  other-modules:       Paths_iris
 123:
     , iris
 136:
  autogen-modules:     Paths_iris
 137:
  other-modules:       Paths_iris
 150:
test-suite iris-test
 160:
    Paths_iris
 164:
    , iris
```

Reply:
Okay, I see the problem. Those functions output line break at the end of line so they are not suitable for this case. I've created a separate issue to implement the required functions:

* https://github.com/chshersh/iris/issues/89

Meanwhile, feel free to keep the version you're the most happy with and we can always update the tutorial later 😌